test-data/unit/check-typeis.test

```diff
@@ -0,0 +1,756 @@
+    pass
+
+def func(name: Iterable[str]):
+    reveal_type(name)  # N: Revealed type is "typing.Iterable[builtins.str]"
+    if is_list_of_str(name):
+        reveal_type(name)  # N: Revealed type is "builtins.list[builtins.str]"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsWithTupleGeneric]
+from typing import TypeVar, Tuple
+from typing_extensions import TypeIs
+
+_T = TypeVar("_T")
+
+def is_two_element_tuple(val: Tuple[_T, ...]) -> TypeIs[Tuple[_T, _T]]:
+    pass
+
+def func(names: Tuple[str, ...]):
+    reveal_type(names)  # N: Revealed type is "builtins.tuple[builtins.str, ...]"
+    if is_two_element_tuple(names):
+        reveal_type(names)  # N: Revealed type is "Tuple[builtins.str, builtins.str]"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsErroneousDefinitionFails]
+from typing_extensions import TypeIs
+
+class Z:
+    def typeis1(self, *, x: object) -> TypeIs[int]:  # E: TypeIs functions must have a positional argument
+        ...
+
+    @staticmethod
+    def typeis2(x: object) -> TypeIs[int]:
+        ...
+
+    @staticmethod
+    def typeis3(*, x: object) -> TypeIs[int]:  # E: TypeIs functions must have a positional argument
+        ...
+
+def bad_typeis(*, x: object) -> TypeIs[int]:  # E: TypeIs functions must have a positional argument
+    ...
+
+[builtins fixtures/classmethod.pyi]
+
+[case testTypeIsWithKeywordArg]
+from typing_extensions import TypeIs
+
+class Z:
+    def typeis(self, x: object) -> TypeIs[int]:
+        ...
+
+def typeis(x: object) -> TypeIs[int]:
+    ...
+
+n: object
+if typeis(x=n):
+    reveal_type(n)  # N: Revealed type is "builtins.int"
+
+if Z().typeis(x=n):
+    reveal_type(n)  # N: Revealed type is "builtins.int"
+[builtins fixtures/tuple.pyi]
+
+[case testStaticMethodTypeIs]
+from typing_extensions import TypeIs
+
+class Y:
+    @staticmethod
+    def typeis(h: object) -> TypeIs[int]:
+        ...
+
+x: object
+if Y().typeis(x):
+    reveal_type(x)  # N: Revealed type is "builtins.int"
+if Y.typeis(x):
+    reveal_type(x)  # N: Revealed type is "builtins.int"
+[builtins fixtures/classmethod.pyi]
+
+[case testTypeIsKwargFollowingThroughOverloaded]
+from typing import overload, Union
+from typing_extensions import TypeIs
+
+@overload
+def typeis(x: object, y: str) -> TypeIs[str]:
+    ...
+
+@overload
+def typeis(x: object, y: int) -> TypeIs[int]:
+    ...
+
+def typeis(x: object, y: Union[int, str]) -> Union[TypeIs[int], TypeIs[str]]:
+    ...
+
+x: object
+if typeis(x=x, y=42):
+    reveal_type(x)  # N: Revealed type is "builtins.int"
+
+if typeis(y=42, x=x):
+    reveal_type(x)  # N: Revealed type is "builtins.int"
+
+if typeis(x=x, y="42"):
+    reveal_type(x)  # N: Revealed type is "builtins.str"
+
+if typeis(y="42", x=x):
+    reveal_type(x)  # N: Revealed type is "builtins.str"
+[builtins fixtures/tuple.pyi]
+
+[case testGenericAliasWithTypeIs]
+from typing import Callable, List, TypeVar
+from typing_extensions import TypeIs
+
+A = Callable[[object], TypeIs[List[T]]]
```

Comment:
This is a general issue with test cases that I've seen in other tests as well. I'm not sure if T is coming from the builtins stubs or whatever else. If you are really interested I could quickly try and search where this has happened.

Obviously still good to point it out and change it here.

Reply:
I think the root cause of this is that a normal `builtins.pyi` in typeshed defined `_T` (which does not get exported) wheres the test fixtures (like `dict.pyi`) typically use stuff like `T` (which does get exported).

It would probably be a good idea to change all fixtures, but that might be quite a bit of work.

Reply:
Added an explicit definition of `T` in this test. Agree that ideally we should fix this throughout the test suite.